### PR TITLE
Support typos and noise modifiers

### LINKED ISSUE #262: Investigate issue with OpusTrainer typos
see https://github.com/hplt-project/OpusTrainer/issues/40

### LINKED ISSUE #348: Leave only 100% augmented datasets for evaluation
From analysis perspective it seems only “-strict” metrics are easy to interpret. They correspond to 100% augmented datasets. Everything else includes some percentages of augmented data and doesn’t tell much. Especially that those percentages that we set in the code are in fact probabilities so it can be inconsistent.

### PR DESCRIPTION
**Pull request by eu9ene**
- update OpusTrainer
- enable typos OpusTrainer modifier
- enable noise OpusTrainer modifier
- remove `strict` modifiers, all single augmentation modifiers now mean 100% augmented data
- add tests
- update docs
- increase validation frequency for the teacher model to stall and end training faster

fixes #262 
fixes #348 

[skip ci]

### CODE REVIEW

**Comment by eu9ene** [comment]
The [experiment](https://firefox-ci-tc.services.mozilla.com/tasks/groups/H396IbuxSAWjhjpBvg_azw) is going well so far and I think it's ready for review. I'm waiting for training to finish to inspect the evaluation results for the augmentations.

**Comment by eu9ene** [comment]
student step is failing: https://firefox-ci-tc.services.mozilla.com/tasks/CcbSB0w_RlKXUmGCkQpr0g/runs/0/logs/public/logs/live.log

I think the issue is when typos change something the alignments stay the same. I'll need to think about this one.

**Comment by eu9ene** [comment]
Ok, I disabled typos for students. For now, let's assume the teacher learns how to handle typos, there are some in the training data and students will adopt this on distillation. I'll think more about it in #439 which includes more work on alignments.

**Comment by marco-c** [comment]
@eu9ene could we add typos for the evaluation part?

**Comment by eu9ene** [comment]
> @eu9ene could we add typos for the evaluation part?

Yes, it's supported, and I'm using it in the current experiment (see [config](https://firefox-ci-tc.services.mozilla.com/tasks/H396IbuxSAWjhjpBvg_azw) `datasets.test` section). Or do you mean the evaluation that we do in the models repo? We can add it there too when we merge this code base in the training repo.

File: pipeline/train/configs/opustrainer/backward.yml [hunk 1]
```diff
@@ -8,10 +8,6 @@ train:
   - original 1.0
   - until original 10 # Train for 10 epochs. Only OpusTrainer can control epochs, it's all one big epoch for Marian
 
-modifiers:
```
[the remark this answers is not in the document]
Comment (gregtatum) [reply pipeline/train/configs/opustrainer/backward.yml:11]: Thinking it through, I don't see why we would augment backwards models.

Suggestion: Can we document in a table in the docs what training and evaluation steps use augmentation? That way it's clear as to our current strategy. Basically anything that takes in a parallel corpus and generates a response.

**Review by gregtatum — APPROVED** [review]
The changes all seem reasonable.

Thought: This isn't using the `run_task` abstraction level, but the unit tests for the python make me reasonably confident in the results. I'm unsure if this is running in the CI training pipeline, but it would be nice, but not required to get coverage at a more integration level.

**Comment by eu9ene** [comment]
I added another change after looking at my experiment that I want to land right away: `valid-freq: 3000` for teacher. This will help to stop training faster. I noticed that our latest experiments are running quite a lot longer than usual. It's because with `valid-freq: 5000` it takes 100k updates to stall 20 times. It used to be 3000 which means 60k to stop with `early-stopping: 20`.

**Pull request merged**